### ISSUE TITLE
AIML chat fails sometimes when it receives "fair enough" as input

### ISSUE THREAD

**Issue by leungmanhin** [opencog/opencog#2365]
it failed with this error
```
ERROR: In procedure opencog-extension:
ERROR: Throw to key `C++-EXCEPTION' with args `("cog-execute!" "Not a link! (/home/leungmanhin/hansonrobotics/opencog/atomspace/opencog/atoms/base/Atom.h:206)")'.
```

it seems that it is related to this psi rule
```
(MemberLink
   (ImplicationLink (stv 1 6.1289907e-06)
      (AndLink
         (AndLink
            (EvaluationLink
               (PredicateNode "*-AIML-pattern-*")
               (ListLink
                  (GlobNode "$star-1")
                  (WordNode "enough" (stv 5.5117674e-05 0.0024937657))
               )
            )
         )
         (ListLink
            (ExecutionOutputLink
               (DefinedSchemaNode "AIML-tag srai")
               (ListLink
                  (ListLink
                     (GlobNode "$star-2")
                  )
               )
            )
            (ExecutionOutputLink
               (DefinedSchemaNode "AIML-tag think")
               (ListLink
                  (ListLink
                     (ExecutionOutputLink
                        (DefinedSchemaNode "AIML-tag set")
                        (ListLink
                           (ConceptNode "topic" (stv 0.0018315018 0.0012484394))
                           (ListLink
                           )
                        )
                     )
                  )
               )
            )
         )
      )
      (ConceptNode "AIML chat subsystem goal")
   )
   (ConceptNode "OpenPsi: AIML chat demand" (stv 0.97000003 1))
)
```

**Comment by leungmanhin** [comment]
I think the MapLink created here: https://github.com/opencog/opencog/blob/master/opencog/nlp/aiml/aiml.scm#L249
```
(MapLink
   (ImplicationLink
      (ListLink
         (GlobNode "$star-1")
         (WordNode "enough" (stv 5.5117674e-05 0.0024937657))
      )
      (ListLink
         (ExecutionOutputLink
            (DefinedSchemaNode "AIML-tag srai")
            (ListLink
               (ListLink
                  (GlobNode "$star-2")
               )
            )
         )
         (ExecutionOutputLink
            (DefinedSchemaNode "AIML-tag think")
            (ListLink
               (ListLink
                  (ExecutionOutputLink
                     (DefinedSchemaNode "AIML-tag set")
                     (ListLink
                        (ConceptNode "topic" (stv 0.0018315018 0.0012484394))
                        (ListLink
                        )
                     )
                  )
               )
            )
         )
      )
   )
   (SetLink
      (ListLink
         (WordNode "[fair]" (stv 2.7558837e-05 0.0012484394))
         (WordNode "enough" (stv 5.5117674e-05 0.0024937657))
      )
   )
)
```
is causing the problem when being executed...

**Comment by linas** [comment]
I can't quite tell, I am guessing that the root cause is a broken AIML rule. the pattern seems to be "* ENOUGH" but there does not seem to be this kind of rule in HEAD -- or I can't find it. Maybe its from character-dev?

The reason for the square brackets around [fair] is that relex (link-grammar) put them there.  This square brackets need to be removed!   This is a fix to relex, I guess.

The generic AIML rules do have a `<pattern>FAIR ENOUGH</pattern>` in them, the square brackets are ruining it.

**Comment by linas** [comment]
this commit to relex fixes the square-bracket issue:
https://github.com/opencog/relex/commit/c945a8729e8562c25287d2e89d90b784f50444ca

**Comment by linas** [comment]
this commit to link-grammar makes "fair enough" parse:  opencog/link-grammar/@1f4c404aa3b4da353337a302bdaa84901a985851

**Comment by linas** [comment]
anyway, the AIML rule still seems broken somehow -- there should not be a star-2 in the pattern, from what I can tell. So the failing AIML rule should be tracked down and fixed. 

In addition, the AIML subsystem needs to be hardend to exceptions, so that if there are invalid AIML rules, we don't crash as a result.

**Comment by leungmanhin** [comment]
Yes I found it in character_dev, indeed it seems to be a broken rule, I've created a PR there, and I am also tracking down all other failing AIML rules that are causing problems...

**Comment by linas** [comment]
Hmm. I am wondering if perhaps `<set name="topic"></set>` is supposed to be valid AIML, whose intent is to clear the current topic (unset the current topic).  The opencog importer expects a valid topic value between the set, here: https://github.com/opencog/opencog/blob/master/opencog/nlp/aiml/import/aiml2psi.pl#L571  so if the goal is to unset the topic, then we need a fix to this, so that if `$3` is empty ... 

Hmm. Actually that's OK. The problem is perhaps here:  https://github.com/opencog/opencog/blob/master/opencog/nlp/aiml/aiml.scm#L495 where the grounded schema node asserts because the empty ListLink means that there is no VALUE....

**Comment by linas** [comment]
Yep, that's it: in the initial bug report, we've got this snippet:
```
                     (ExecutionOutputLink
                        (DefinedSchemaNode "AIML-tag set")
                        (ListLink
                           (ConceptNode "topic" (stv 0.0018315018 0.0012484394))
                           (ListLink
                           )
                        )
```
which is OK, but then this causes this to run: https://github.com/opencog/opencog/blob/master/opencog/nlp/aiml/aiml.scm#L498 which causes flat-val to not be an atom, which causes the exception ... or something like that. 

So, I guess that ... I dunno, I'm not sure .. if the VALUE is an empty ListLink, then the StateLink needs to be set to ... ?what?   Should the StateLink be removed? Or should the state be set to the empty word .. i.e. set the state to `(WordNode "")`

**Comment by linas** [comment]
So I think that maybe #2372 fixes this ... @leungmanhin can you test and confirm or deny and then merge, as needed?

**Comment by linas** [comment]
Arghh... but there is also a different bug, too.  The line that says this:
```
(ExecutionOutputLink
               (DefinedSchemaNode "AIML-tag srai")
               (ListLink
                  (ListLink
                     (GlobNode "$star-2")
                  )
               )
            )
```
should say star-1 not star-2.   

I am extremely short on time, and might not be able to get around to looking at this for many days. I will try tonight, but not sure...

**Comment by linas** [comment]
Oh gosh . I wonder it the strung-out star in the bad rule is causing all star numbering to be wrong?
The problem is with these:
```
<category>
￼  <pattern>Do you have a favorite poem</pattern>
￼  <template><srai>*</srai></template>
</category>
<category>
   <pattern> * enough </pattern>
   <template><srai>*</srai><think><set name="topic"></set></think></template>
</category>
```

**Comment by linas** [comment]
OK, so #2373 looks like a major fix to how AIML is getting handled.  It fixes the star-counting problem above. I can't merge it right now, cause I must run an errand before 5PM today, and I'm worried this change might have other negative impacts -- the perl code is a bit of spaghetti, so I'm not sure of what other side effects there might be....

**Comment by leungmanhin** [comment]
for some reason I cannot get it to execute the above rule again because it fails at `is-usable-rule?` (https://github.com/opencog/opencog/blob/master/opencog/nlp/aiml/aiml.scm#L230)...

**Comment by leungmanhin** [comment]
never mind, my mistake

**Comment by leungmanhin** [comment]
Fixed by #2373
